app/src/main/java/com/sakaimobile/development/sakaiclient20/ui/activities/SitePageActivity.java

```diff
@@ -0,0 +1,136 @@
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Set;
+
+
+public class SitePageActivity extends AppCompatActivity {
+
+
+    protected Set<LiveData> beingObserved;
+
+    @Override
+    protected void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        setContentView(R.layout.activity_site_page);
+
+        beingObserved = new HashSet<>();
+
+        Intent i = getIntent();
+        String siteType = i.getStringExtra(getString(R.string.site_type_tag));
+        Course course = (Course) i.getSerializableExtra(getString(R.string.course_tag));
+
+        // setup toolbar
+        Toolbar toolbar = findViewById(R.id.toolbar);
+        setSupportActionBar(toolbar);
+        toolbar.setNavigationOnClickListener(v -> this.onBackPressed());
+
+
+        // set the toolbar title as siteType + coursename
+        getSupportActionBar().setTitle(String.format("%s: %s", siteType, course.title));
+
+
+
+        // load the appropriate fragment for the site type
+        if (siteType.equals(getString(R.string.gradebook))) {
+            startSiteGradesFragment(course);
+        } else if (siteType.equals(getString(R.string.announcements_site))) {
+            startSiteAnnouncementsFragment(course);
+        } else if (siteType.equals(getString(R.string.resources_site))) {
+            startSiteResourcesFragment(course);
+        } else {
+            startWebViewFragment(siteType, course);
+        }
+    }
+
+
+    private void startSiteGradesFragment(Course course) {
+        Bundle bun = new Bundle();
+        bun.putString(getString(R.string.siteid_tag), course.siteId);
+
+        SiteGradesFragment fragment = new SiteGradesFragment();
+        fragment.setArguments(bun);
+
+        addFragment(fragment);
+    }
+
+    private void startSiteAnnouncementsFragment(Course course) {
+
+        HashMap<String, Course> siteIdToCourse = new HashMap<>();
+        siteIdToCourse.put(course.siteId, course);
+
+        Bundle bun = new Bundle();
+        bun.putString(getString(R.string.siteid_tag), course.siteId);
+        bun.putSerializable(getString(R.string.siteid_to_course_map), siteIdToCourse);
+
+        AnnouncementsFragment fragment = new AnnouncementsFragment();
+        fragment.setArguments(bun);
+
+        addFragment(fragment);
+    }
+
+    private void startSiteResourcesFragment(Course course) {
+        Bundle bun = new Bundle();
+        bun.putString(getString(R.string.siteid_tag), course.siteId);
+
+        SiteResourcesFragment fragment = new SiteResourcesFragment();
+        fragment.setArguments(bun);
+
+        addFragment(fragment);
+    }
+
+    private void startWebViewFragment(String siteName, Course course) {
+
+        String url = null;
+        for(SitePage page : course.sitePages) {
+            if(page.title.equals(siteName))
+                url = page.url;
+        }
+
+        if(url != null) {
```

Comment:
Is the URL ever null?

Reply:
it shouldn't be, but its just a failsafe, should I remove the null check?